openpype/hosts/houdini/plugins/publish/collect_husk_rop.py

```diff
@@ -0,0 +1,178 @@
+            render_products = []
+
+            # Default beauty AOV
+            beauty_product = self.get_render_product_name(
+                prefix=default_prefix, suffix=None
+            )
+            render_products.append(beauty_product)
+
+            files_by_aov = {
+                "beauty": self.generate_expected_files(instance,
+                                                       beauty_product)
+            }
+
+        for product in render_products:
+            self.log.debug("Found render product: %s" % product)
+
+        filenames = list(render_products)
+        instance.data["files"] = filenames
+        instance.data["renderProducts"] = colorspace.ARenderProduct()
+
+        # For now by default do NOT try to publish the rendered output
+        instance.data["publishJobState"] = "Suspended"
+        instance.data["attachTo"] = []      # stub required data
+
+        if "expectedFiles" not in instance.data:
+            instance.data["expectedFiles"] = list()
+        instance.data["expectedFiles"].append(files_by_aov)
+
+        # update the colorspace data
+        colorspace_data = get_color_management_preferences()
+        instance.data["colorspaceConfig"] = colorspace_data["config"]
+        instance.data["colorspaceDisplay"] = colorspace_data["display"]
+        instance.data["colorspaceView"] = colorspace_data["view"]
+
+        # Generate the commandline
+        instance.data["huskCommandline"] = self._generate_command(rop,
+                                                             instance)
+
+    def get_render_product_name(self, prefix, suffix):
+        product_name = prefix
+        if suffix:
+            # Add ".{suffix}" before the extension
+            prefix_base, ext = os.path.splitext(prefix)
+            product_name = prefix_base + "." + suffix + ext
+
+        return product_name
+
+    def generate_expected_files(self, instance, path):
+        """Create expected files in instance data"""
+
+        dir = os.path.dirname(path)
+        file = os.path.basename(path)
+
+        if "#" in file:
+            def replace(match):
+                return "%0{}d".format(len(match.group()))
+
+            file = re.sub("#+", replace, file)
+
+        if "%" not in file:
+            return path
+
+        expected_files = []
+        start = instance.data["frameStart"]
+        end = instance.data["frameEnd"]
+        for i in range(int(start), (int(end) + 1)):
+            expected_files.append(
+                os.path.join(dir, (file % i)).replace("\\", "/"))
+
+        return expected_files
+
+    def _generate_command(self, node, instance):
+        chunkSize = node.parm('chunkSize').eval()
+        renderer = node.parm('renderer')
+        self.log.debug("renderer: %s" % renderer)
+        verbosity = node.parm('verbosity').eval()
+
+        # hardcoded usdfile --  we must replace but how?
+        usdfile = "the_path_for_usdfile.usd"
+        self.log.debug("usdfile: %s" % usdfile)
```

Comment:
FYI, The USD file that the USD Render ROP will generate is defined by parms on the node.

See:
![image](https://github.com/ynput/OpenPype/assets/2439881/2d6eafab-0ea2-473d-b0ff-a566018fd980)

Note that you can also disable the **Render** of the USD render rop:
![image](https://github.com/ynput/OpenPype/assets/2439881/c31a7d25-0437-40b2-9767-c1337a27e6a1)

I believe if you disable the Render Command it will only write out the USD file itself. As such then the ROP node outputs what you'd want (create USD file; no render) - and it's only up to the Husk command to pick up the file.

So likely what we'd want is to update the this:
![image](https://github.com/ynput/OpenPype/assets/2439881/9ff595d3-ad16-40aa-9b4a-0eb9f4c0ea44)

To a path that is accessible by all farm machines. If we e.g. set it on the submission to deadline then we also know what output 'render' file it will generate and we can directly define the accurate server path for the USD file that is then accessible for all machines on the farm.

@spmhickstein is that what you're looking for or might have been missing? Admittedly - I haven't read too much of your code yet so you might already be doing all of this.

Reply:
> the ideal would be a cached file and version tracked from AYON Publish, this is usually the workflow we follow.

It doesn't have to be 'version tracked AYON publish' I'd say - to me it's more a transient/intermediate render file for that particular render session.

Note the the USD render rop by default is set to this:
![image](https://github.com/ynput/OpenPype/assets/2439881/24e2b144-7fd6-40a2-941e-0dbfa168c70a)

So that if you change the USD file save path to say:

```
$HIP/render_files/$HIPNAME/__render__.usd
```

Then it'd basically write into a folder `render_files` subfolder next to the workfiles, generating a unique render cache folder for that particular hip file. The render farm machines then just pick up that **non-temporary** file.

The USD saving _can_ write additional files (like in memory referenced layers) depending on the settings - that's why I'm writing into a `$HIPNAME` subfolder to try to be on the safe side.

What do you think?